src/WiFiManager.cpp

```diff
@@ -1,228 +1,206 @@
-  ESP_LOGD(TAG, "Scan completed");
-
-  std::uint16_t numNetworks = WiFi.scanComplete();
-  if (numNetworks < 0) {
-    if (numNetworks != WIFI_SCAN_RUNNING) {
-      ESP_LOGE(TAG, "Scan failed");
-      SetWiFiState(WiFiState::Disconnected);
-      CaptivePortal::Start();
-    } else {
-      ESP_LOGE(TAG, "Scan is still running");
-    }
-    return;
-  }
-
-  DynamicJsonDocument doc(64 + numNetworks * 128);
-  JsonArray networks = doc.createNestedArray("networks");
-
-  if (numNetworks == 0) {
-    ESP_LOGD(TAG, "No networks found");
-    SetWiFiState(WiFiState::Disconnected);
-    CaptivePortal::Start();
-    CaptivePortal::BroadcastMessageJSON(doc);
-    return;
-  }
-
-  for (auto& cred : s_wifiCredentials) {
-    cred.wifiIndex = UINT16_MAX;
-  }
-
-  std::uint16_t recognizedNetworks = 0;
-  for (std::uint16_t i = 0; i < numNetworks; i++) {
-    String ssid = WiFi.SSID(i);
-    bool saved  = false;
-    for (auto& cred : s_wifiCredentials) {
-      if (cred.ssid == ssid) {
-        cred.wifiIndex = i;
-        recognizedNetworks++;
-        saved = true;
-        break;
-      }
-    }
-    JsonObject network = networks.createNestedObject();
-    network["index"]   = i;
-    network["ssid"]    = ssid;
-    network["bssid"]   = WiFi.BSSIDstr(i);
-    network["rssi"]    = WiFi.RSSI(i);
-    network["channel"] = WiFi.channel(i);
-    network["saved"]   = saved;
-  }
-
-  CaptivePortal::BroadcastMessageJSON(doc);
-
-  if (recognizedNetworks == 0) {
-    ESP_LOGD(TAG, "No recognized networks found");
-    SetWiFiState(WiFiState::Disconnected);
-    CaptivePortal::Start();
-    return;
-  }
-
-  // Attempt to connect to the first recognized network
-  for (auto& cred : s_wifiCredentials) {
-    if (cred.wifiIndex != UINT16_MAX) {
-      ESP_LOGD(TAG, "Attempting to connect to %s", cred.ssid.c_str());
-      WiFi.begin(cred.ssid.c_str(), cred.password.c_str());
-      SetWiFiState(WiFiState::Connecting);
-      return;
-    }
-  }
-}
 void _evWiFiConnected(arduino_event_id_t event, arduino_event_info_t info) {
   ESP_LOGD(TAG, "WiFi connected");
-  SetWiFiState(WiFiState::Connected);
+  OpenShock::SetWiFiState(WiFiState::Connected);
   CaptivePortal::Stop();
 }
 void _evWiFiDisconnected(arduino_event_id_t event, arduino_event_info_t info) {
   ESP_LOGD(TAG, "WiFi disconnected");
-  SetWiFiState(WiFiState::Disconnected);
+  OpenShock::SetWiFiState(WiFiState::Disconnected);
   CaptivePortal::Start();
 }
+void _evWiFiNetworkDiscovered(const wifi_ap_record_t* record) {
+  WiFiNetwork network {
+    .ssid              = {0},
+    .bssid             = {0},
+    .channel           = record->primary,
+    .rssi              = record->rssi,
+    .authmode          = record->authmode,
+    .reconnectionCount = 0,
+    .credentialsId     = UINT8_MAX,
+  };
+
+  static_assert(sizeof(network.ssid) == sizeof(record->ssid), "SSID size mismatch");
+  memcpy(network.ssid, record->ssid, sizeof(network.ssid));
```

Comment:
I see `memcpy` but do you not forget to set the terminating null byte here?

Reply:
memcpy is different from strcpy, strcpy respects null terminator and does some extra unneccesary checks that we dont need because we know both ssid types are the same size, and we also know that the ssid we are copying from has a valid null terminator